### PR TITLE
ContextMenu: fix menu height on async render (T1258881)

### PR DESCRIPTION
**Pull request by anna-shakhova**

### CODE REVIEW

**Comment by anna-shakhova** [comment]
[ContextMenu - Incorrect initial menu height in React 18 [T1258881]](https://trello.com/c/yIeFA8Pz/9479-contextmenu-incorrect-initial-menu-height-in-react-18-t1258881)

File: packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts [hunk 1]
```diff
@@ -531,6 +531,7 @@ class ContextMenu extends MenuBase {
 
   _renderSubmenuItems(node, $itemFrame: dxElementWrapper): void {
     this._renderItems(this._getChildNodes(node), $itemFrame);
+    this._planPostRenderActions();
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: I see that this function is called on render in `async` strategy, so now it's called twice? Doesn't it have side affects?
It's more expected for me if we call it once after everything is rendered

File: packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts [hunk 2]
```diff
@@ -1130,6 +1133,10 @@ class ContextMenu extends MenuBase {
   hide(): Promise<unknown> {
     return this.toggle(false);
   }
+
+  _postProcessRenderItems() {
+    this._dimensionChanged(false);
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: Looks not very logical that we call `dimensionChanged` after all items are rendered. Cannot we split the code another way? Like extract necessary code from `_dimensionChanged` to a new method and reuse it here

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js [hunk 3]
```diff
@@ -60,5 +67,59 @@ QUnit.module('Context menu', () => {
...
+        instance.option('onShown', (e) => {
+            const $submenus = $(`.${DX_SUBMENU_CLASS}`);
+            const $nestedSubmenu = $submenus.eq(1);
+            const $nestedSubmenuItemsContainer = $nestedSubmenu.find(`.${DX_CONTEXT_MENU_ITEMS_CONTAINER_CLASS}`);
+            const $scrollableContainer = $nestedSubmenu.find(`.${DX_SCROLLABLE_CONTENT_CLASS}`);
+
+            assert.roughEqual($nestedSubmenuItemsContainer.outerHeight(), $scrollableContainer.outerHeight(), .1);
+            done();
+        });
+
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]:
```suggestion
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js [hunk 4]
```diff
@@ -4,6 +4,13 @@ import ContextMenu from 'ui/context_menu';
 import 'ui/button';
 import 'generic_light.css!';
 
+const DX_SUBMENU_CLASS = 'dx-submenu';
+const DX_CONTEXT_MENU_ITEMS_CONTAINER_CLASS = 'dx-menu-items-container';
+const DX_SCROLLABLE_CONTENT_CLASS = 'dx-scrollable-container';
+const DX_MENU_ITEM_CLASS = 'dx-menu-item';
+
+const ITEMS_CONTAINER_PADDING = 1;
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]: Now it's not used in the code

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js [hunk 5]
```diff
@@ -60,5 +67,59 @@ QUnit.module('Context menu', () => {
...
+                    }
+                }
+            },
+        });
+
+        $(menuTargetSelector).trigger($.Event('dxcontextmenu'));
+
+        const $itemsContainer = instance.itemsContainer();
+        const $rootItem = $itemsContainer.find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+        $($itemsContainer).trigger($.Event('dxhoverstart', { target: $rootItem.get(0) }));
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]: hover cannot be used on mobile devices. Feel free to do smth like
```javascript
        if(!isDeviceDesktop(assert)) {
            assert.ok(true, 'not actual for mobile devices');
            return;
        }
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js (hunk 3 above)
Comment (anna-shakhova) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js (hunk 4 above)
Comment (anna-shakhova) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]: thanks, removed

File: packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts (hunk 1 above)
Comment (anna-shakhova) [reply packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: unfortunately, submenu is not contextMenu instance, and this method not called for it, so we have to call it manually after all submenu items are rendered

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js (hunk 5 above)
Comment (anna-shakhova) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.async.tests.js]: added click event for mobile devices

File: packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts (hunk 2 above)
Comment (anna-shakhova) [reply packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: changed logic - now we do not change height after render, instead height calculated and submenu visiblility changed yo visible after render is done

File: packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts [hunk 6]
```diff
@@ -704,7 +704,7 @@ class ContextMenu extends MenuBase {
     return availableHeight - SUBMENU_PADDING;
   }
 
-  _dimensionChanged() {
+  _dimensionChanged(showAnimation = true) {
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: Cannot these changed be reverted now?

Comment (anna-shakhova) [reply packages/devextreme/js/__internal/ui/context_menu/m_context_menu.ts]: thanks, forgot to remove

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Pull request merged**